src/core/qgstaskmanager.cpp

```diff
@@ -142,6 +142,27 @@ QList<QgsMapLayer *> QgsTask::dependentLayers() const
   return _qgis_listQPointerToRaw( mDependentLayers );
 }
 
+bool QgsTask::waitForFinished( int timeout )
```

Comment:
There is one open question here:
Right now it uses a QEventLoop internally. This is dangerous (no need to rediscuss `processEvents()`). I wonder if it's possible to replace this with a different approach that doesn't run the event loop. The problem is, that this will probably freeze if the task is still queued because the task manager depends on events (statusChanged signals from tasks to trigger `processQueue()`)...

Maybe this needs to be (dummy code)

    QgsTaskManager::waitForFinished( QgsTask* task, int timeout )
    {
      if ( !task->isRunning() )
      {
         someOtherTask->pause();
         task->continue();
      }
      
      QMutex mutex;
      task->setMutex( mutex );
      mutex.tryLock( timeout );
    }